cmd/src/actions_exec_backend_runner.go

```diff
@@ -363,3 +359,62 @@ func repositoryZipArchiveURL(repoName, rev, token string) (*url.URL, error) {
 	u.Path = path.Join(u.Path, repoName+"@"+rev, "-", "raw")
 	return u, nil
 }
+
+func unzip(zipFile, dest string) error {
+	r, err := zip.OpenReader(zipFile)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+
+	for _, f := range r.File {
+		fpath := filepath.Join(dest, f.Name)
+
+		// Check for ZipSlip. More Info: http://bit.ly/2MsjAWE
+		if !strings.HasPrefix(fpath, filepath.Clean(dest)+string(os.PathSeparator)) {
+			return fmt.Errorf("%s: illegal file path", fpath)
+		}
+
+		if f.FileInfo().IsDir() {
+			err := os.MkdirAll(fpath, os.ModePerm)
+			if err != nil {
+				return err
+			}
+			continue
+		}
+
+		if err := os.MkdirAll(filepath.Dir(fpath), os.ModePerm); err != nil {
+			return err
+		}
+
+		outFile, err := os.OpenFile(fpath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, f.Mode())
+		if err != nil {
+			return err
+		}
+
+		rc, err := f.Open()
+		if err != nil {
+			outFile.Close()
+			return err
+		}
+
+		if _, err = io.Copy(outFile, rc); err != nil {
```

Comment:
I would recommend doing this differently. As written, we're still missing a close of `outFile` if `rc.Close` fails. And we actually don't _care_ if closing `rc` fails, because it's only open for reading so there is no risk to data. At best we might want to log that.

What I would write here is:

```go
_, err := io.Copy(outFile, rc)
rc.Close()
_, cerr := outFile.Close()
// Now we have safely closed everything that needs it, and can check errors
if err != nil {
   return fmt.Errorf("copying %q: %v", f.Name, err)
}
if cerr != nil {
  return fmt.Errorf("closing: %v", cerr)
}
```

I know this pattern seems a little weird at first, but it's a lot easier to be sure we haven't missed a case if we do error checking as a separate phase. We _always_ want to close, regardless of what fails, and only after we've done that do we care whether something broke.

Reply:
Ah, _now_ I get it. Thank you!